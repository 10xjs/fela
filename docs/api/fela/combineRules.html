
<!DOCTYPE HTML>
<html lang="en" >
    <head>
        <title>5.1.3. combineRules Â· Fela</title>
        <meta charset="UTF-8">
        <meta http-equiv="X-UA-Compatible" content="IE=edge" />
        <meta content="text/html; charset=utf-8" http-equiv="Content-Type">
        <meta name="description" content="">
        <meta name="generator" content="GitBook 3.1.0">
        <meta name="author" content="Robin Frischmann <robin@rofrischmann.de>">
        
        
    
    <link rel="stylesheet" href="../../../gitbook/style.css">

    
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-prism/prism.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-advanced-emoji/emoji-website.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-search/search.css">
                
            
                
                <link rel="stylesheet" href="../../../gitbook/gitbook-plugin-fontsettings/website.css">
                
            
        

    

    
        
    
        
    
        
    
        
    
        
    
        
    

        
    
    
    <meta name="HandheldFriendly" content="true"/>
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    <link rel="apple-touch-icon-precomposed" sizes="152x152" href="../../../gitbook/images/apple-touch-icon-precomposed-152.png">
    <link rel="shortcut icon" href="../../../gitbook/images/favicon.ico" type="image/x-icon">

    
    <link rel="next" href="enhance.html" />
    
    
    <link rel="prev" href="Renderer.html" />
    

    </head>
    <body>
        
<div class="book">
    <div class="book-summary">
        
            
<div id="book-search-input" role="search">
    <input type="text" placeholder="Type to search" />
</div>

            
                <nav role="navigation">
                


<ul class="summary">
    
    

    

    
        
        
    
        <li class="chapter " data-level="1.1" data-path="../../../">
            
                <a href="../../../">
            
                    
                    Read Me
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2" data-path="../../Introduction.html">
            
                <a href="../../Introduction.html">
            
                    
                    1. Introduction
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.2.1" data-path="../../introduction/Motivation.html">
            
                <a href="../../introduction/Motivation.html">
            
                    
                    1.1. Motivation
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.2" data-path="../../introduction/Principles.html">
            
                <a href="../../introduction/Principles.html">
            
                    
                    1.2. Principles
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.3" data-path="../../introduction/Benefits.html">
            
                <a href="../../introduction/Benefits.html">
            
                    
                    1.3. Benefits
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.4" data-path="../../introduction/Drawbacks.html">
            
                <a href="../../introduction/Drawbacks.html">
            
                    
                    1.4. Drawbacks
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.5" data-path="../../introduction/Ecosystem.html">
            
                <a href="../../introduction/Ecosystem.html">
            
                    
                    1.5. Ecosystem
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.2.6" data-path="../../introduction/Examples.html">
            
                <a href="../../introduction/Examples.html">
            
                    
                    1.6. Examples
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.3" data-path="../../Basics.html">
            
                <a href="../../Basics.html">
            
                    
                    2. Basics
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.3.1" data-path="../../basics/Rules.html">
            
                <a href="../../basics/Rules.html">
            
                    
                    2.1. Rules
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.2" data-path="../../basics/Keyframes.html">
            
                <a href="../../basics/Keyframes.html">
            
                    
                    2.2. Keyframes
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.3" data-path="../../basics/Fonts.html">
            
                <a href="../../basics/Fonts.html">
            
                    
                    2.3. Fonts
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.3.4" data-path="../../basics/Renderer.html">
            
                <a href="../../basics/Renderer.html">
            
                    
                    2.4. Renderer
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.4" data-path="../../Advanced.html">
            
                <a href="../../Advanced.html">
            
                    
                    3. Advanced
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.4.1" data-path="../../advanced/DOMRendering.html">
            
                <a href="../../advanced/DOMRendering.html">
            
                    
                    3.1. DOM Rendering
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.2" data-path="../../advanced/ServerRendering.html">
            
                <a href="../../advanced/ServerRendering.html">
            
                    
                    3.2. Server Rendering
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.3" data-path="../../advanced/Plugins.html">
            
                <a href="../../advanced/Plugins.html">
            
                    
                    3.3. Plugins
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.4" data-path="../../advanced/RendererConfiguration.html">
            
                <a href="../../advanced/RendererConfiguration.html">
            
                    
                    3.4. Renderer Configuration
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.5" data-path="../../advanced/StaticStyle.html">
            
                <a href="../../advanced/StaticStyle.html">
            
                    
                    3.5. Global & Third-Party Styles
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.6" data-path="../../advanced/CombinedRules.html">
            
                <a href="../../advanced/CombinedRules.html">
            
                    
                    3.6. Combined Rules
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.7" data-path="../../advanced/Enhancers.html">
            
                <a href="../../advanced/Enhancers.html">
            
                    
                    3.7. Enhancers
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.4.8" data-path="../../advanced/UnderTheHood.md">
            
                <span>
            
                    
                    3.7. Under The Hood
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.5" data-path="../../UsageGuides.html">
            
                <a href="../../UsageGuides.html">
            
                    
                    4. Usage Guides
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.5.1" data-path="../../guides/UsageWithReact.html">
            
                <a href="../../guides/UsageWithReact.html">
            
                    
                    4.1. Usage with React
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.2" data-path="../../guides/UsageWithReactNative.html">
            
                <a href="../../guides/UsageWithReactNative.html">
            
                    
                    4.2. Usage with React Native
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.3" data-path="../../guides/UsageWithAngular2.html">
            
                <a href="../../guides/UsageWithAngular2.html">
            
                    
                    4.3. Usage with Angular 2
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.5.4" data-path="../../guides/UsageWithInferno.html">
            
                <a href="../../guides/UsageWithInferno.html">
            
                    
                    4.4. Usage with Inferno
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6" data-path="../../API.html">
            
                <a href="../../API.html">
            
                    
                    5. API Reference
            
                </a>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1" >
            
                <span>
            
                    
                    5.1. fela
            
                </span>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.1.1" data-path="createRenderer.html">
            
                <a href="createRenderer.html">
            
                    
                    5.1.1. createRenderer
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.1.2" data-path="Renderer.html">
            
                <a href="Renderer.html">
            
                    
                    5.1.2. Renderer
            
                </a>
            

            
        </li>
    
        <li class="chapter active" data-level="1.6.1.3" data-path="combineRules.html">
            
                <a href="combineRules.html">
            
                    
                    5.1.3. combineRules
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.6.1.4" data-path="enhance.html">
            
                <a href="enhance.html">
            
                    
                    5.1.4. enhance
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.6.2" >
            
                <span>
            
                    
                    5.2. fela-dom
            
                </span>
            

            
            <ul class="articles">
                
    
        <li class="chapter " data-level="1.6.2.1" data-path="../fela-dom/render.html">
            
                <a href="../fela-dom/render.html">
            
                    
                    5.2.1. render
            
                </a>
            

            
        </li>
    

            </ul>
            
        </li>
    

            </ul>
            
        </li>
    
        <li class="chapter " data-level="1.7" data-path="../../Troubleshooting.html">
            
                <a href="../../Troubleshooting.html">
            
                    
                    6. Troubleshooting
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.8" data-path="../../FAQ.html">
            
                <a href="../../FAQ.html">
            
                    
                    7. FAQ
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.9" data-path="../../Feedback.html">
            
                <a href="../../Feedback.html">
            
                    
                    8. Feedback
            
                </a>
            

            
        </li>
    
        <li class="chapter " data-level="1.10" data-path="../../Thanks.html">
            
                <a href="../../Thanks.html">
            
                    
                    9. Thanks
            
                </a>
            

            
        </li>
    

    

    <li class="divider"></li>

    <li>
        <a href="https://www.gitbook.com" target="blank" class="gitbook-link">
            Published with GitBook
        </a>
    </li>
</ul>


                </nav>
            
        
    </div>

    <div class="book-body">
        
            <div class="body-inner">
                
                    

<div class="book-header" role="navigation">
    

    <!-- Title -->
    <h1>
        <i class="fa fa-circle-o-notch fa-spin"></i>
        <a href="../../.." >5.1.3. combineRules</a>
    </h1>
</div>




                    <div class="page-wrapper" tabindex="-1" role="main">
                        <div class="page-inner">
                            
<div id="book-search-results">
    <div class="search-noresults">
    
                                <section class="normal markdown-section">
                                
                                <h1 id="combinerulesrules"><code>combineRules(...rules)</code></h1>
<p>A common pattern when styling elements is to have some base styles which get applied to every instance as well as some extended styles which get added based on some condition. Just concatenating classNames may lead to specificity problems as the last rendered rule always wins.<br>
To solve this issue, <code>combineRules</code> allows you to compose multiple rules into a single super selector.</p>
<h2 id="arguments">Arguments</h2>
<ol>
<li><code>...rules</code> (<em>arguments</em>): Functions returning valid <a href="../../basics/Rules.html#styleobject">style objects</a>.</li>
</ol>
<h2 id="returns">Returns</h2>
<p>(<em>Function</em>) A super selector which composes all <code>rules</code> from left to right.</p>
<h2 id="example">Example</h2>
<pre class="language-"><code class="lang-javascript"><span class="token keyword">import</span> <span class="token punctuation">{</span> combineRules <span class="token punctuation">}</span> <span class="token keyword">from</span> <span class="token string">&apos;fela&apos;</span>

<span class="token keyword">const</span> renderer <span class="token operator">=</span> <span class="token function">createRenderer</span><span class="token punctuation">(</span>mountNode<span class="token punctuation">)</span>

<span class="token keyword">const</span> rule <span class="token operator">=</span> props <span class="token operator">=</span><span class="token operator">&gt;</span> <span class="token punctuation">(</span><span class="token punctuation">{</span>
  fontSize<span class="token punctuation">:</span> props<span class="token punctuation">.</span>fontSize
  color<span class="token punctuation">:</span> <span class="token string">&apos;red&apos;</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>

<span class="token keyword">const</span> anotherRule <span class="token operator">=</span> props <span class="token operator">=</span><span class="token operator">&gt;</span> <span class="token punctuation">(</span><span class="token punctuation">{</span>
  color<span class="token punctuation">:</span> <span class="token string">&apos;blue&apos;</span>
<span class="token punctuation">}</span><span class="token punctuation">)</span>

<span class="token keyword">const</span> superRule <span class="token operator">=</span> <span class="token function">combineRule</span><span class="token punctuation">(</span>rule<span class="token punctuation">,</span> anotherRule<span class="token punctuation">)</span>

renderer<span class="token punctuation">.</span><span class="token function">render</span><span class="token punctuation">(</span>rule<span class="token punctuation">,</span> <span class="token punctuation">{</span> fontSize<span class="token punctuation">:</span> <span class="token string">&apos;12px &apos;</span><span class="token punctuation">}</span><span class="token punctuation">)</span> <span class="token comment" spellcheck="true">// =&gt; c1 c2</span>
renderer<span class="token punctuation">.</span><span class="token function">render</span><span class="token punctuation">(</span>anotherRule<span class="token punctuation">)</span> <span class="token comment" spellcheck="true">// =&gt; c3</span>
renderer<span class="token punctuation">.</span><span class="token function">render</span><span class="token punctuation">(</span>superRule<span class="token punctuation">,</span> <span class="token punctuation">{</span> fontSize<span class="token punctuation">:</span> <span class="token string">&apos;12px&apos;</span> <span class="token punctuation">}</span><span class="token punctuation">)</span> <span class="token comment" spellcheck="true">// c4 c5</span>

console<span class="token punctuation">.</span><span class="token function">log</span><span class="token punctuation">(</span>renderer<span class="token punctuation">.</span><span class="token function">renderToString</span><span class="token punctuation">(</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
<span class="token comment" spellcheck="true">// .c1{color:red}.c2{font-size:12px}</span>
<span class="token comment" spellcheck="true">// .c3{color:blue}</span>
<span class="token comment" spellcheck="true">// .c4{color:blue}.c5{font-size:12px}</span>
</code></pre>
<h2 id="tips">Tips</h2>
<ul>
<li>Try to avoid <code>combineRules</code> if possible. If you are using <code>combineRules</code> excessively you might have to rethink and refactor your styling technique to achieve a more modular and simple one.</li>
<li><code>combineRules</code> will always create a new rule which prevents using already cached sub-rules. Therefore it might negatively affect the rendering performance if you are using multiple combined rules.</li>
</ul>

                                
                                </section>
                            
    </div>
    <div class="search-results">
        <div class="has-results">
            
            <h1 class="search-results-title"><span class='search-results-count'></span> results matching "<span class='search-query'></span>"</h1>
            <ul class="search-results-list"></ul>
            
        </div>
        <div class="no-results">
            
            <h1 class="search-results-title">No results matching "<span class='search-query'></span>"</h1>
            
        </div>
    </div>
</div>

                        </div>
                    </div>
                
            </div>

            
                
                <a href="Renderer.html" class="navigation navigation-prev " aria-label="Previous page: 5.1.2. Renderer">
                    <i class="fa fa-angle-left"></i>
                </a>
                
                
                <a href="enhance.html" class="navigation navigation-next " aria-label="Next page: 5.1.4. enhance">
                    <i class="fa fa-angle-right"></i>
                </a>
                
            
        
    </div>

    <script>
        var gitbook = gitbook || [];
        gitbook.push(function() {
            gitbook.page.hasChanged({"page":{"title":"5.1.3. combineRules","level":"1.6.1.3","depth":3,"next":{"title":"5.1.4. enhance","level":"1.6.1.4","depth":3,"path":"docs/api/fela/enhance.md","ref":"docs/api/fela/enhance.md","articles":[]},"previous":{"title":"5.1.2. Renderer","level":"1.6.1.2","depth":3,"path":"docs/api/fela/Renderer.md","ref":"docs/api/fela/Renderer.md","articles":[]},"dir":"ltr"},"config":{"plugins":["edit-link","prism","-highlight","github","anker-enable","advanced-emoji"],"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"pluginsConfig":{"prism":{},"github":{"url":"https://github.com/rofrischmann/fela/"},"anker-enable":{},"search":{},"lunr":{"maxIndexSize":1000000},"fontsettings":{"theme":"white","family":"sans","size":2},"advanced-emoji":{"embedEmojis":false},"sharing":{"facebook":true,"twitter":true,"google":false,"weibo":false,"instapaper":false,"vk":false,"all":["facebook","google","twitter","weibo","instapaper"]},"edit-link":{"base":"https://github.com/rofrischmann/fela/tree/master","label":"Edit This Page"},"theme-default":{"styles":{"website":"styles/website.css","pdf":"styles/pdf.css","epub":"styles/epub.css","mobi":"styles/mobi.css","ebook":"styles/ebook.css","print":"styles/print.css"},"showLevel":false}},"theme":"default","author":"Robin Frischmann <robin@rofrischmann.de>","pdf":{"pageNumbers":true,"fontSize":12,"fontFamily":"Arial","paperSize":"a4","chapterMark":"pagebreak","pageBreaksBefore":"/","margin":{"right":62,"left":62,"top":56,"bottom":56}},"structure":{"langs":"LANGS.md","readme":"README.md","glossary":"GLOSSARY.md","summary":"SUMMARY.md"},"variables":{},"title":"Fela","language":"en","gitbook":"3.1.0","description":"Official documentation for Fela"},"file":{"path":"docs/api/fela/combineRules.md","mtime":"2016-12-11T15:25:08.000Z","type":"markdown"},"gitbook":{"version":"3.1.0","time":"2016-12-11T21:43:13.007Z"},"basePath":"../../..","book":{"language":""}});
        });
    </script>
</div>

        
    <script src="../../../gitbook/gitbook.js"></script>
    <script src="../../../gitbook/theme.js"></script>
    
        
        <script src="../../../gitbook/gitbook-plugin-edit-link/plugin.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-github/plugin.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-anker-enable/anker.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-search/search-engine.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-search/search.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-lunr/lunr.min.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-lunr/search-lunr.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-sharing/buttons.js"></script>
        
    
        
        <script src="../../../gitbook/gitbook-plugin-fontsettings/fontsettings.js"></script>
        
    

    </body>
</html>

