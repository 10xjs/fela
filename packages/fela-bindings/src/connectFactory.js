/* @flow */
import reduce from 'lodash/reduce'

import generateDisplayName from './generateDisplayName'
import hoistStatics from './hoistStatics'

export default function connectFactory(
  BaseComponent: any,
  createElement: Function,
  withTheme: Function,
  contextTypes?: Object
): Function {
  return function connect(rules: Object | Function): Function {
    return (component: any): any => {
      class EnhancedComponent extends BaseComponent {
        static displayName = generateDisplayName(component)

        render() {
          const { renderer } = this.context

          const preparedRules =
            typeof rules === 'function' ? rules(this.props) : rules

          const styles = reduce(
            preparedRules,
            (styleMap, rule, name) => {
              const preparedRule =
                typeof rule !== 'function' ? () => rule : rule
              styleMap[name] = renderer.renderRule(preparedRule, this.props)
              return styleMap
            },
            {}
          )

          const { theme, ...propsWithoutTheme } = this.props
          return createElement(component, {
            ...propsWithoutTheme,
            styles
          })
        }
      }

      if (contextTypes) {
        EnhancedComponent.contextTypes = contextTypes
      }

      const themedComponent = withTheme(EnhancedComponent)
      return hoistStatics(themedComponent, component)
    }
  }
}
