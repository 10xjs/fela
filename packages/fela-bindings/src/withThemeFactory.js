/* @flow */
import hoistStatics from './hoistStatics'

export default function withThemeFactory(
  createElement: Function,
  FelaTheme: Function
): Function {
  return function withTheme(
    component: Object,
    propName?: string = 'theme'
  ): Object {
    const WithTheme = props =>
      createElement(FelaTheme, undefined, theme =>
        createElement(component, {
          ...props,
          [propName]: theme,
        })
      )

    return hoistStatics(WithTheme, component)
  }
}
