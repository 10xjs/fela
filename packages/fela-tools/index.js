module.exports = require('fela/lib/tools/index.js')
