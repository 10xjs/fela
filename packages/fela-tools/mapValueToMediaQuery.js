module.exports = require('fela/lib/tools/mapValueToMediaQuery')
