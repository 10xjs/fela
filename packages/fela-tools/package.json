{
  "name": "fela-tools",
  "version": "4.2.3",
  "description": "Useful tools for working with Fela",
  "main": "index.js",
  "files": [
    "LICENSE",
    "README.md",
    "index.js",
    "StyleSheet.js",
    "mapValueToMediaQuery.js",
    "/dist"
  ],
  "repository": "https://github.com/rofrischmann/fela/",
  "keywords": [
    "fela",
    "fela-tool",
    "cssinjs"
  ],
  "author": "Robin Frischmann",
  "license": "MIT",
  "peerDependencies": {
    "fela": "4.2.3"
  }
}