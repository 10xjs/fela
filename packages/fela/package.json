{
  "name": "fela",
  "version": "6.0.4",
  "description": "Style as a Function of State",
  "typings": "index.d.ts",
  "main": "lib/index.js",
  "module": "es/index.js",
  "jsnext:main": "es/index.js",
  "files": [
    "LICENSE",
    "README.md",
    "index.d.ts",
    "lib/**",
    "es/**"
  ],
  "repository": "https://github.com/rofrischmann/fela/",
  "keywords": [
    "fela",
    "dynamic styling",
    "stylesheet",
    "css",
    "styling",
    "functional styling",
    "functional css",
    "atomic css",
    "atomic design",
    "cssinjs"
  ],
  "author": "Robin Frischmann",
  "license": "MIT",
  "dependencies": {
    "css-in-js-utils": "2.0.0",
    "fela-utils": "^7.0.4"
  },
  "devDependencies": {
    "fela-tools": "^5.0.12"
  }
}
