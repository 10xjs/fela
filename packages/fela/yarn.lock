# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
# yarn lockfile v1


css-in-js-utils@^3.0.0:
  version "3.0.0"
  resolved "https://registry.yarnpkg.com/css-in-js-utils/-/css-in-js-utils-3.0.0.tgz#4379185f5cc79f9eba39b4e795c317f253ffaa40"
  dependencies:
    hyphenate-style-name "^1.0.2"

csstype@^2.5.5:
  version "2.5.7"
  resolved "https://registry.yarnpkg.com/csstype/-/csstype-2.5.7.tgz#bf9235d5872141eccfb2d16d82993c6b149179ff"

fast-loops@^1.0.0:
  version "1.0.1"
  resolved "https://registry.yarnpkg.com/fast-loops/-/fast-loops-1.0.1.tgz#a3d33e71654a7e6e67d7c2de0fee4ea11c6f2526"

hyphenate-style-name@^1.0.2:
  version "1.0.2"
  resolved "https://registry.yarnpkg.com/hyphenate-style-name/-/hyphenate-style-name-1.0.2.tgz#31160a36930adaf1fc04c6074f7eb41465d4ec4b"

isobject@^3.0.1:
  version "3.0.1"
  resolved "https://registry.yarnpkg.com/isobject/-/isobject-3.0.1.tgz#4e431e92b11a9731636aa1f9c8d1ccbcfdab78df"
