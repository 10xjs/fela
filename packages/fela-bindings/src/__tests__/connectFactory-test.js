import React, { createElement, Component } from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { html as beautify } from 'js-beautify'

import { renderToString } from 'fela-tools'
import { createRenderer } from 'fela'

import connectFactory from '../connectFactory'
import withThemeFactory from '../withThemeFactory'

const withTheme = withThemeFactory(Component, createElement, {
  theme: PropTypes.object
})

const connect = connectFactory(Component, createElement, withTheme, {
  renderer: PropTypes.object
})

describe('Connect Factory for bindings', () => {
  it('should process rules and create classNames', () => {
    const rules = {
      rule1: () => ({
        padding: 1
      }),
      rule2: props => ({
        color: props.color
      })
    }

    const MyComponent = connect(rules)(({ styles }) => (
      <div>
        <span className={styles.rule1} />
        <span className={styles.rule2} />
      </div>
    ))

    MyComponent.defaultProps = {
      color: 'red'
    }

    const renderer = createRenderer()
    const wrapper = mount(<MyComponent />, {
      context: {
        renderer
      }
    })

    expect([
      beautify(`<style>${renderToString(renderer)}</style>`),
      toJson(wrapper)
    ]).toMatchSnapshot()
  })
})
