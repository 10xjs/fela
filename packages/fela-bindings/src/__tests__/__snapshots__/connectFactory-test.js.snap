// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connect Factory for bindings should component receive rules prop with all combined rules 1`] = `
Object {
  "css": ".a {
    padding: 2
}

.b {
    font-size: 16
}

.c {
    color: red
}",
  "html": "<div><div><span class=\\"a\\"></span><span class=\\"c b\\"></span></div></div>",
}
`;

exports[`Connect Factory for bindings should compose styles 1`] = `
Object {
  "css": ".a {
    padding: 2
}

.b {
    color: red
}

.c {
    font-size: 16
}",
  "html": "<div><span class=\\"a\\"></span><span class=\\"b c\\"></span></div>",
}
`;

exports[`Connect Factory for bindings should extend the rule properties 1`] = `
Object {
  "css": ".a {
    padding: 2
}

.b {
    color: red
}

.c {
    font-size: 16
}",
  "html": "<div><span class=\\"a\\"></span><span class=\\"b c\\"></span></div>",
}
`;

exports[`Connect Factory for bindings should not pass through "theme" prop when used without "ThemeProvider" 1`] = `
Object {
  "css": ".a {
    padding: 1
}

.b {
    color: red
}",
  "html": "<div color=\\"red\\"><span class=\\"a\\"></span><span class=\\"b\\"></span></div>",
}
`;

exports[`Connect Factory for bindings should process rules and create classNames 1`] = `
Object {
  "css": ".a {
    padding: 1
}

.b {
    color: red
}",
  "html": "<div><span class=\\"a\\"></span><span class=\\"b\\"></span></div>",
}
`;

exports[`Connect Factory for bindings should process rules and create classNames with rules as function 1`] = `
Object {
  "css": ".a {
    padding: 1
}

.b {
    color: red
}",
  "html": "<div><span class=\\"a\\"></span><span class=\\"b\\"></span></div>",
}
`;

exports[`Connect Factory for bindings should provide rules prop for connected component which is an object with rules in the values of fields 1`] = `
Object {
  "css": ".a {
    padding: 1
}

.b {
    color: red
}",
  "html": "",
}
`;
